rfcs/0020-project_api_extend_metadata.md

```diff
@@ -0,0 +1,252 @@
+
+However that's a problem when the API user (e.g. TVMC) needs to automatically
+determine the options available for the API methods, like when automatically
+building a command line parser with subcommand domains that closely mapped to
+the API methods (e.g. subcommands to create, build, and flash a project).
+
+Moreover, currently it's impossible to determine which option is required and
+which one is optional, so it would be at least necessary for the API user to
+build a static ad hoc table with all options available on a given project
+stating which option is required and which one is optional for the project. This
+is impractical to maintain and would result in the API user having to update the
+the static table every time an option is added, removed, or modified in the
+Project API server.
+
+Hence to ease the automatic detection of the options available on each Project
+API method the following two new metadata are proposed: `required` and
+`optional`.
+
+Both will contain a list of method names for which the option is available,
+either as a required option (if in `required` list), or as an optional option
+(if in `optional` list). At least one API method must be listed in `required` or
+in the `optional` list. A method name must be listed only in the `required` or
+in the `optional` list, i.e. an option can't be required and optional at the
+same time for given API method. An option can be required for a method and
+optional for another method.
+
+The elements in the lists `required` and `optional` must be in the set of method
+names implemented by the ProjectAPIClient class and that have the parameter
+`options` defined. These methods are dispatched to the server, which implements
+the server counterparts to properly handle the client dispatches and
+ultimately defines the options available for each API method. The current method
+name that satisfy these criteria are `generate_project`, `build`, `flash`, and
+`open_transport`.
+
+`required` metadatum or `optional` metadatum (or both) must be provided for
+every option.
+
+## On "type" metadatum
+
+The option type can sometimes be determined implicitly by what's return
+in metadatum `choices`, but this not ideal. For example, for option `verbose` it
+would be possible to infer it is a boolean option and therefore it can be
+converted to a command line flag if metadatum `choices` is a couple of True and
+False. Nonetheless that would lead to cumbersome logic at API user side (e.g.
+TVMC) to infer the option type, like iterating over the tuple elements to search
+for True or False. This can be solved directly if the option type is returned
+explicitly with the option.
+
+Thus adding a `type` metadatum allows a much simpler way for the API users to
+determine the type of an option when that is necessary for various reasons, like
+when building a command line parser based on the available project options.
+
+The following types, passed as strings, are proposed for the `type` metadatum:
```

Comment:
can you state the rationale behind these limits to capture it here? IIRC we settled on only non-complex JSON-serializable types.

Reply:
Sure. Do you think it's correct to use **packed** types to refer to dict, tuple, and list instead of complex  types (the non-complex types would be the non-packed types so)?

Reply:
do you mean `struct.pack` or `PackedFunc`? I don't think `PackedFunc` has a place in this RPC, as it's using JSON-RPC not TVM RPC. I would think we would use JS objects to represent more complex values if we did go that direction, but honestly i prefer to not if possible to avoid over-complicating this.

Reply:
@areusch I didn't mean neither `struct.pack` nor `PackedFunc`. :) I'm actually acknowledging that we've settled on only non-complex JSON-serialized types. My last question is almost a nit, and I just would like to know if it would be ok to use the term 'packed' types instead of 'non-complex' types because I think Python uses that term to refer to built-in types like dict, tuple, list etc -- which we would like to not use here as we have settled.